File: lantern/src/main/java/com/github/nisrulz/lantern/Lantern.java
/*
 * Copyright (C) 2017 Nishant Srivastava
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.nisrulz.lantern;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.os.Handler;

import java.lang.ref.WeakReference;
import java.util.concurrent.TimeUnit;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresPermission;
import androidx.lifecycle.Lifecycle.Event;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.OnLifecycleEvent;

import static com.github.nisrulz.lantern.Utils.isMarshmallowAndAbove;

public class Lantern implements LifecycleObserver {

    @Nullable
    private final Context context;
    private final Handler handler;
    private final Utils utils;
    private WeakReference<Activity> activityWeakRef;
    @Nullable
    private DisplayLightController displayLightController;
    @Nullable
    private FlashController flashController;
    private long pulseTime = 1000;
    private final Runnable pulseRunnable = new Runnable() {
        @Override
        public void run() {
            if (flashController != null) {
                enableTorchMode(!flashController.torchEnabled());
                handler.postDelayed(pulseRunnable, pulseTime);
            }
        }
    };

    public Lantern(@Nullable Context context) {
        this.context = context;
        utils = new Utils();
        handler = new Handler();
    }

    public Lantern setupDisplayController(Activity activity) {
        this.activityWeakRef = new WeakReference<>(activity);
        if (displayLightController == null) {
            displayLightController = new DisplayLightControllerImpl(getActivityRef());
        }
        return this;
    }

    public Lantern alwaysOnDisplay(boolean enabled) {
        if (displayLightController != null) {
            if (enabled) {

                displayLightController.enableAlwaysOnMode();
            } else {
                displayLightController.disableAlwaysOnMode();
            }
        }
        return this;
    }

    public Lantern autoBright(boolean enabled) {
        if (displayLightController != null) {
            if (enabled) {
                displayLightController.enableAutoBrightMode();
            } else {
                displayLightController.disableAutoBrightMode();
            }
        }
        return this;
    }

    public Lantern checkAndRequestSystemPermission() {
        if (displayLightController != null) {
            displayLightController.requestSystemWritePermission();
        }
        return this;
    }

    public boolean isSystemWritePermissionGranted() {
        boolean result = false;

        if (displayLightController != null) {
            result = displayLightController.checkSystemWritePermission();
        }

        return result;
    }

    @OnLifecycleEvent(Event.ON_DESTROY)
    public void cleanup() {
        handler.removeCallbacks(pulseRunnable);
        if (displayLightController != null) {
            displayLightController.cleanup();
        }
        this.activityWeakRef = null;
    }

    public Lantern enableTorchMode(boolean enabled) {
        if (flashController != null) {
            if (context != null) {
                if (enabled) {
                    if (!flashController.torchEnabled()
                            && utils.checkForCameraPermission(context.getApplicationContext())) {
                        flashController.on();
                    }
                } else {
                    if (flashController.torchEnabled()
                            && utils.checkForCameraPermission(context.getApplicationContext())) {
                        flashController.off();
                    }
                }
            } else {
                flashController.off();
            }
        }
        return this;
    }

    public Lantern fullBrightDisplay(boolean enabled) {
        if (displayLightController != null) {
            if (enabled) {
                displayLightController.enableFullBrightMode();
            } else {
                displayLightController.disableFullBrightMode();
            }
        }
        return this;
    }

    @RequiresPermission(Manifest.permission.CAMERA)
    public boolean initTorch() {
        if (context != null) {
            if (utils.checkIfCameraFeatureExists(context)
                    && utils.checkForCameraPermission(context)) {
                if (isMarshmallowAndAbove()) {
                    flashController = new PostMarshmallow(context);
                } else {
                    flashController = new PreMarshmallow();
                }
                return true;
            }
        }
        return false;
    }

    public Lantern observeLifecycle(LifecycleOwner lifecycleOwner) {
        // Subscribe to listening lifecycle
        lifecycleOwner.getLifecycle().addObserver(this);

        return this;
    }

    public Lantern pulse(boolean enabled) {
        if (enabled) {
            handler.postDelayed(pulseRunnable, pulseTime);
        } else {
            handler.removeCallbacks(pulseRunnable);
        }
        return this;
    }

    public Lantern withDelay(long time, final TimeUnit timeUnit) {
        this.pulseTime = TimeUnit.MILLISECONDS.convert(time, timeUnit);
        return this;
    }

    /**
     * Torch Enabled State
     *
     * @return boolean
     */
    @RequiresPermission(Manifest.permission.CAMERA)
    public boolean isTorchEnabled() {
        if (initTorch() && flashController != null) {
            return flashController.torchEnabled();
        }

        return false;
    }

    private Activity getActivityRef() {
        return activityWeakRef.get();
    }
}

File: lantern/src/main/java/com/github/nisrulz/lantern/Lantern.kt
/*
 * Copyright (C) 2017 Nishant Srivastava
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.nisrulz.lantern

import android.Manifest
import android.app.Activity
import android.content.Context
import android.os.Handler
import androidx.annotation.Nullable
import androidx.annotation.RequiresPermission
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent
import java.lang.ref.WeakReference
import java.util.concurrent.TimeUnit

class Lantern(private val context: Context?) : LifecycleObserver {
    private val handler: Handler = Handler()
    private val utils: Utils = Utils()
    private var activityWeakRef: WeakReference<Activity>? = null

    @Nullable
    private var displayLightController: DisplayLightController? = null

    @Nullable
    private var flashController: FlashController? = null
    private var pulseTime: Long = 1000
    private val pulseRunnable: Runnable = object : Runnable {
        override fun run() {
            if (flashController != null) {
                enableTorchMode(flashController?.torchEnabled()?.not() == true)
                handler.postDelayed(this, pulseTime)
            }
        }
    }

    fun setupDisplayController(activity: Activity): Lantern {
        activityWeakRef = WeakReference<Activity>(activity)

        activityRef?.let {
            if (displayLightController == null) {
                displayLightController = DisplayLightControllerImpl(it)
            }
        }

        return this
    }

    fun alwaysOnDisplay(enabled: Boolean): Lantern {
        if (displayLightController != null) {
            if (enabled) {
                displayLightController?.enableAlwaysOnMode()
            } else {
                displayLightController?.disableAlwaysOnMode()
            }
        }
        return this
    }

    fun autoBright(enabled: Boolean): Lantern {
        if (displayLightController != null) {
            if (enabled) {
                displayLightController?.enableAutoBrightMode()
            } else {
                displayLightController?.disableAutoBrightMode()
            }
        }
        return this
    }

    fun checkAndRequestSystemPermission(): Lantern {
        if (displayLightController != null) {
            displayLightController?.requestSystemWritePermission()
        }
        return this
    }

    val isSystemWritePermissionGranted: Boolean
        get() {
            var result = false
            if (displayLightController != null) {
                result = displayLightController?.checkSystemWritePermission() == true
            }
            return result
        }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun cleanup() {
        handler.removeCallbacks(pulseRunnable)
        if (displayLightController != null) {
            displayLightController?.cleanup()
        }
        activityWeakRef = null
    }

    fun enableTorchMode(enabled: Boolean): Lantern {
        if (flashController != null) {
            if (context != null) {
                if (enabled) {
                    if (flashController?.torchEnabled()?.not() == true
                        && utils.checkForCameraPermission(context.applicationContext)
                    ) {
                        flashController?.on()
                    }
                } else {
                    if (flashController?.torchEnabled() == true
                        && utils.checkForCameraPermission(context.applicationContext)
                    ) {
                        flashController?.off()
                    }
                }
            } else {
                flashController?.off()
            }
        }
        return this
    }

    fun fullBrightDisplay(enabled: Boolean): Lantern {
        if (displayLightController != null) {
            if (enabled) {
                displayLightController?.enableFullBrightMode()
            } else {
                displayLightController?.disableFullBrightMode()
            }
        }
        return this
    }

    @RequiresPermission(Manifest.permission.CAMERA)
    fun initTorch(): Boolean {
        if (context != null) {
            if (utils.checkIfCameraFeatureExists(context)
                && utils.checkForCameraPermission(context)
            ) {
                flashController = if (Utils.isMarshmallowAndAbove) {
                    PostMarshmallow(context)
                } else {
                    PreMarshmallow()
                }
                return true
            }
        }
        return false
    }

    fun observeLifecycle(lifecycleOwner: LifecycleOwner): Lantern {
        // Subscribe to listening lifecycle
        lifecycleOwner.lifecycle.addObserver(this)
        return this
    }

    fun pulse(enabled: Boolean): Lantern {
        if (enabled) {
            handler.postDelayed(pulseRunnable, pulseTime)
        } else {
            handler.removeCallbacks(pulseRunnable)
        }
        return this
    }

    fun withDelay(time: Long, timeUnit: TimeUnit?): Lantern {
        pulseTime = TimeUnit.MILLISECONDS.convert(time, timeUnit)
        return this
    }

    /**
     * Torch Enabled State
     *
     * @return boolean
     */
    @get:RequiresPermission(Manifest.permission.CAMERA)
    val isTorchEnabled: Boolean
        get() = if (initTorch() && flashController != null) {
            flashController?.torchEnabled() ?: false
        } else false
    private val activityRef: Activity?
        get() = activityWeakRef?.get()
}